api/types.go

```diff
@@ -0,0 +1,141 @@
+package api
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
+)
+
+// LoadBalancerList is a collection of LoadBalancer
+type LoadBalancerList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LoadBalancer `json:"items"`
+}
+
+// +genclient=true
+
+// LoadBalancer describe a LoadBalancer witch providers Load balancing for services
+type LoadBalancer struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of the LoadBalancer
+	LoadBalancerSpec `json:"spec,omitempty"`
+}
+
+// LoadBalancerSpec is a description of a LoadBalancer
+type LoadBalancerSpec struct {
+	// Type determines the type of LoadBalancer
+	// if type is internal, the providers MUST and ONLY can be service
+	Type LoadBalancerType `json:"type"`
+	// Replica only used when Provider's type is service
+	// you can not use replica and nodes at the same time
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+	// Specification of the desired behavior of the nodes
+	// If you set nodes, replicas will be disabled
+	// +optional
+	Nodes *NodesSpec `json:"nodes,omitempty"`
+	// Specification of the desired behavior of the proxy
+	Proxy ProxySpec `json:"proxy"`
+	// Specification of the desired behavior of the providers
+	Providers ProvidersSpec `json:"providers"`
+}
+
+// LoadBalancerType ...
+type LoadBalancerType string
+
+const (
+	// LoadBalancerTypeInternal is internal type of LoadBalancer
+	LoadBalancerTypeInternal = "internal"
+	// LoadBalancerTypeExternal is internal type of LoadBalancer
+	LoadBalancerTypeExternal = "external"
+)
+
+// NodesSpec is a description of nodes
+type NodesSpec struct {
+	Names []string `json:"names"`
+	// +optional
+	Dedicated *apiv1.TaintEffect `json:"dedicated,omitempty"`
```

Comment:
I'm not a fan of this name, this should either be:

```go
Dedicated *bool `json:"dedicated,omitempty"`
```

Where depending on the value, controller decides what taints to use, e.g. if dedicated=true, then use NoSchedule, or false do nothing.

If we want user to provide the taint effect, then i'd suggest rename to Effect,

```go
Effect *apiv1.TaintEffect `json:"dedicated,omitempty"`
```

Using bool is easier to use but less flexible, but Effect requires more learning.  I don't have a big preference, but generally, people would go for flexibility, so let's do a rename.